File: backend/src/docs/docs.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { DocumentationEntry } from '../common/entities/documentation-entry.entity';
import { CreateDocDto } from './dto/create-doc.dto';
import { UpdateDocDto } from './dto/update-doc.dto';

@Injectable()
export class DocsService {
  constructor(
    @InjectRepository(DocumentationEntry)
    private readonly docsRepository: Repository<DocumentationEntry>
  ) {}

  async list(params?: {
    menuSlug?: string;
    category?: string;
    enabled?: boolean;
    search?: string;
  }) {
    const qb = this.docsRepository.createQueryBuilder('doc');

    if (params?.menuSlug) {
      qb.andWhere('doc.menuSlug = :menuSlug', {
        menuSlug: this.normalizeSlug(params.menuSlug)
      });
    }
    if (params?.category) {
      qb.andWhere('doc.category = :category', {
        category: this.normalizeSlug(params.category)
      });
    }
    if (typeof params?.enabled === 'boolean') {
      qb.andWhere('doc.enabled = :enabled', { enabled: params.enabled });
    }
    if (params?.search) {
      qb.andWhere('(doc.title LIKE :q OR doc.content LIKE :q)', {
        q: `%${params.search}%`
      });
    }

    return qb
      .orderBy('doc.category', 'ASC')
      .addOrderBy('doc.orderIndex', 'ASC')
      .addOrderBy('doc.createdAt', 'ASC')
      .getMany();
  }

  async create(dto: CreateDocDto) {
    const entry = this.docsRepository.create({
      menuSlug: this.normalizeSlug(dto.menuSlug),
      category: this.normalizeCategory(dto.category),
      title: dto.title,
      content: dto.content,
      link: dto.link ?? null,
      orderIndex: dto.orderIndex ?? 0,
      enabled: dto.enabled ?? true
    });
    return this.docsRepository.save(entry);
  }

  async getById(id: string) {
    const entry = await this.docsRepository.findOne({ where: { id } });
    if (!entry) {
      throw new NotFoundException('Documentation entry not found');
    }
    return entry;
  }

  async update(id: string, dto: UpdateDocDto) {
    const entry = await this.docsRepository.findOne({ where: { id } });
    if (!entry) {
      throw new NotFoundException('Documentation entry not found');
    }

    Object.assign(entry, dto);
    if (dto.menuSlug) {
      entry.menuSlug = this.normalizeSlug(dto.menuSlug);
    }
    if (dto.category) {
      entry.category = this.normalizeCategory(dto.category);
    }

    return this.docsRepository.save(entry);
  }

  async remove(id: string) {
    const entry = await this.docsRepository.findOne({ where: { id } });
    if (!entry) {
      throw new NotFoundException('Documentation entry not found');
    }
    await this.docsRepository.remove(entry);
    return { id };
  }

  private normalizeSlug(value: string) {
    return value.trim().toLowerCase();
  }

  private normalizeCategory(value?: string) {
    if (!value) {
      return 'general';
    }
    return this.normalizeSlug(value);
  }
}

File: backend-java/src/main/java/com/neria/manager/docs/DocsService.java
package com.neria.manager.docs;

import com.neria.manager.common.entities.DocumentationEntry;
import com.neria.manager.common.repos.DocumentationEntryRepository;
import java.time.LocalDateTime;
import java.util.List;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

@Service
public class DocsService {
  private final DocumentationEntryRepository repository;

  public DocsService(DocumentationEntryRepository repository) {
    this.repository = repository;
  }

  public List<DocumentationEntry> list(String menuSlug, String category, Boolean enabled, String search) {
    String normalizedMenu = normalizeSlug(menuSlug);
    String normalizedCategory = normalizeSlug(category);
    if (normalizedMenu == null && normalizedCategory == null && enabled == null && (search == null || search.isBlank())) {
      return repository.findAllByOrderByCategoryAscOrderIndexAscCreatedAtAsc();
    }
    return repository.search(normalizedMenu, normalizedCategory, enabled, search != null ? search.trim() : null);
  }

  public DocumentationEntry getById(String id) {
    return repository.findById(id)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Documentation entry not found"));
  }

  public DocumentationEntry create(CreateDocRequest dto) {
    DocumentationEntry entry = new DocumentationEntry();
    entry.setId(java.util.UUID.randomUUID().toString());
    entry.setMenuSlug(normalizeSlug(dto.menuSlug));
    entry.setCategory(normalizeCategory(dto.category));
    entry.setTitle(dto.title);
    entry.setContent(dto.content);
    entry.setLink(dto.link);
    entry.setOrderIndex(dto.orderIndex != null ? dto.orderIndex : 0);
    entry.setEnabled(dto.enabled != null ? dto.enabled : true);
    entry.setCreatedAt(LocalDateTime.now());
    entry.setUpdatedAt(LocalDateTime.now());
    return repository.save(entry);
  }

  public DocumentationEntry update(String id, UpdateDocRequest dto) {
    DocumentationEntry entry = repository.findById(id)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Documentation entry not found"));

    if (dto.menuSlug != null) {
      entry.setMenuSlug(normalizeSlug(dto.menuSlug));
    }
    if (dto.category != null) {
      entry.setCategory(normalizeCategory(dto.category));
    }
    if (dto.title != null) {
      entry.setTitle(dto.title);
    }
    if (dto.content != null) {
      entry.setContent(dto.content);
    }
    if (dto.link != null) {
      entry.setLink(dto.link);
    }
    if (dto.orderIndex != null) {
      entry.setOrderIndex(dto.orderIndex);
    }
    if (dto.enabled != null) {
      entry.setEnabled(dto.enabled);
    }
    entry.setUpdatedAt(LocalDateTime.now());
    return repository.save(entry);
  }

  public java.util.Map<String, Object> remove(String id) {
    DocumentationEntry entry = repository.findById(id)
        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Documentation entry not found"));
    repository.delete(entry);
    return java.util.Map.of("id", id);
  }

  private String normalizeSlug(String value) {
    if (value == null) {
      return null;
    }
    return value.trim().toLowerCase();
  }

  private String normalizeCategory(String value) {
    if (value == null || value.isBlank()) {
      return "general";
    }
    return normalizeSlug(value);
  }

  public static class CreateDocRequest {
    public String menuSlug;
    public String category;
    public String title;
    public String content;
    public String link;
    public Integer orderIndex;
    public Boolean enabled;
  }

  public static class UpdateDocRequest {
    public String menuSlug;
    public String category;
    public String title;
    public String content;
    public String link;
    public Integer orderIndex;
    public Boolean enabled;
  }
}
